go/vt/vtgate/executor.go

```diff
@@ -284,24 +300,38 @@ func (e *Executor) handleExec(ctx context.Context, safeSession *SafeSession, sql
 	return qr, err
 }
 
-func (e *Executor) shardExec(ctx context.Context, safeSession *SafeSession, sql string, bindVars map[string]*querypb.BindVariable, target querypb.Target, logStats *LogStats) (*sqltypes.Result, error) {
+func (e *Executor) destinationExec(ctx context.Context, safeSession *SafeSession, sql string, bindVars map[string]*querypb.BindVariable, target querypb.Target, destination key.Destination, logStats *LogStats) (*sqltypes.Result, error) {
 	f := func() ([]*srvtopo.ResolvedShard, error) {
-		return e.resolver.resolver.ResolveDestination(ctx, target.Keyspace, target.TabletType, key.DestinationShard(target.Shard))
+		var result []*srvtopo.ResolvedShard
+		var err error
+		result, err = e.resolver.resolver.ResolveDestination(ctx, target.Keyspace, target.TabletType, destination)
+		if err != nil {
+			return nil, err
+		}
+		logStats.ShardQueries = uint32(len(result))
+		return result, nil
 	}
-	return e.resolver.Execute(ctx, sql, bindVars, target.Keyspace, target.TabletType, safeSession.Session, f, false /* notInTransaction */, safeSession.Options, logStats)
+	result, err := e.resolver.Execute(ctx, sql, bindVars, target.TabletType, safeSession.Session, f, false /* notInTransaction */, safeSession.Options, logStats)
+	return result, err
 }
 
 func (e *Executor) handleDDL(ctx context.Context, safeSession *SafeSession, sql string, bindVars map[string]*querypb.BindVariable, target querypb.Target, logStats *LogStats) (*sqltypes.Result, error) {
 	if target.Keyspace == "" {
 		return nil, errNoKeyspace
 	}
+	keyRange, err := parseRange(safeSession.TargetString)
+	if err != nil {
+		return nil, errNoKeyspace
+	}
 
 	f := func() ([]*srvtopo.ResolvedShard, error) {
 		var destination key.Destination
```

Comment:
These six lines can be taken out of the closure, and right after the previous block:
keyRange, err := ...
...
var destination key.Destination
if keyRange != nil {
...

f := func() (....)

Reply:
Done. When I did this, I also notice that then this can also re-use `e.destinationExec`